### PR TITLE
8252847: Optimize primitive arrayCopy stubs using AVX-512 masked instructions

### PR DESCRIPTION
**Pull request by jatin-bhateja**
Summary:

1)  New AVX3 optimized stubs for both conjoint and disjoint arraycopy.
2)  Special instruction sequence blocks for copy sizes b/w 32-192 bytes.
3)  Block copy operation above 192 bytes is performed using destination address aligned PRE-MAIN-POST loop. Main loop copies 192 byte in one iteration and tail part fall over special instruction sequence blocks.
4)  Both small copy block and aligned loop use 32 byte vector register to prevent and frequency penalty for copy sizes less than AVX3Threshold.
5)  For block size above AVX3Theshold both special blocks and loop operate using 64 byte register.
6)  In case user sets the maximum vector size to 32 bytes, forward copy (disjoint) operations are done using efficient REP MOVS for copy sizes above 4096 bytes.

JMH Results:
  System     :  CascadeLake Server, Intel(R) Xeon(R) Platinum 8280L CPU @ 2.70GHz
  Micros     :  test/micro/org/openjdk/bench/java/lang/ArrayCopy*.java
  Baseline   :  [http://cr.openjdk.java.net/~jbhateja/8252847/JMH_results/ArrayCopy_AVX3_Stubs_Baseline.txt]()
  WithOpt  :  [http://cr.openjdk.java.net/~jbhateja/8252847/JMH_results/ArrayCopy_AVX3_Stubs_WithOpts.txt]()

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252847](https://bugs.openjdk.java.net/browse/JDK-8252847): Optimize primitive arrayCopy stubs using AVX-512 masked instructions 

### Reviewers
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**) ⚠️ Review applies to 78c4fe731653c5e87cd313b432ad7d22b0dac5b9
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/61/head:pull/61`
`$ git checkout pull/61`

### CODE REVIEW

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 1]
```diff
@@ -7963,6 +7963,77 @@ void MacroAssembler::cache_wbsync(bool is_pre)
     sfence();
   }
 }
+
+void MacroAssembler::copy64_masked_avx(Register dst, Register src, XMMRegister xmm,
+                                       KRegister mask, Register length, Register temp,
+                                       BasicType type, int offset, bool use64byteVector) {
+  assert(MaxVectorSize >= 32, "vector length < 32");
+  use64byteVector |= MaxVectorSize > 32 && AVX3Threshold == 0;
+  if (use64byteVector == false) {
```
Comment (neliasso) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:7972]: Change to "!use64byteVector"

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 2]
```diff
@@ -7963,6 +7963,77 @@ void MacroAssembler::cache_wbsync(bool is_pre)
...
+  assert(MaxVectorSize >= 32, "vector length < 32");
+  vmovdqu(xmm, Address(src, offset));
+  vmovdqu(Address(dst, offset), xmm);
+}
+
+
+void MacroAssembler::copy64_avx(Register dst, Register src, XMMRegister xmm, int offset, bool use64byteVector) {
+  assert(MaxVectorSize == 64 || MaxVectorSize == 32, "vector length mismatch");
+  use64byteVector |= MaxVectorSize > 32 && AVX3Threshold == 0;
+  if (use64byteVector == false) {
```
Comment (neliasso) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:8012]: Change to "!use64byteVector"

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 3]
```diff
@@ -7963,6 +7963,77 @@ void MacroAssembler::cache_wbsync(bool is_pre)
...
+  } else {
+     evmovdquq(xmm, Address(src, offset), Assembler::AVX_512bit);
+     evmovdquq(Address(dst, offset), xmm, Assembler::AVX_512bit);
+  }
+}
+
+void MacroAssembler::copy64_conjoint_avx(Register dst, Register src, XMMRegister xmm, int offset, bool use64byteVector) {
+  assert(MaxVectorSize == 64 || MaxVectorSize == 32, "vector length mismatch");
+  use64byteVector |= MaxVectorSize > 32 && AVX3Threshold == 0;
+  if (use64byteVector == false) {
```
Comment (neliasso) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:8026]: Change to "!use64byteVector"

File: src/hotspot/cpu/x86/vm_version_x86.cpp [hunk 4]
```diff
@@ -1162,7 +1164,7 @@ void VM_Version::get_processor_features() {
 #endif // COMPILER2 && ASSERT
 
   if (!FLAG_IS_DEFAULT(AVX3Threshold)) {
-    if (!is_power_of_2(AVX3Threshold)) {
+    if (AVX3Threshold !=0 && !is_power_of_2(AVX3Threshold)) {
```
Comment (neliasso) [inline src/hotspot/cpu/x86/vm_version_x86.cpp:1167]: Missing space before '0'

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 5]
```diff
@@ -7963,6 +7963,77 @@ void MacroAssembler::cache_wbsync(bool is_pre)
     sfence();
   }
 }
+
+void MacroAssembler::copy64_masked_avx(Register dst, Register src, XMMRegister xmm,
+                                       KRegister mask, Register length, Register temp,
+                                       BasicType type, int offset, bool use64byteVector) {
+  assert(MaxVectorSize >= 32, "vector length < 32");
```
Comment (neliasso) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:7970]: Why does "MaxVectorSize >= 32" imply that "vector length < 32"?

This assert appears in multiple locations.

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 6]
```diff
@@ -7963,6 +7963,77 @@ void MacroAssembler::cache_wbsync(bool is_pre)
...
+  use64byteVector |= MaxVectorSize > 32 && AVX3Threshold == 0;
```
Comment (neliasso) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:7971]: When do you expect AVX3Threshold to be 0?

**Review by neliasso — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp (hunk 6 above)
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp:7971]: As of now when user explicitly pass -XX:AVX3Threshold=0 , default value of AVX3Threshold is 4096.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 7]
```diff
@@ -1884,6 +2387,10 @@ class StubGenerator: public StubCodeGenerator {
   address generate_conjoint_int_oop_copy(bool aligned, bool is_oop, address nooverlap_target,
                                          address *entry, const char *name,
                                          bool dest_uninitialized = false) {
+    if (VM_Version::supports_avx512vlbw() && false == is_oop && MaxVectorSize  >= 32) {
```
Comment (neliasso) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2390]: "false == is_oop"  => !is_oop

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 8]
```diff
@@ -1991,6 +2498,9 @@ class StubGenerator: public StubCodeGenerator {
   //
   address generate_disjoint_long_oop_copy(bool aligned, bool is_oop, address *entry,
                                           const char *name, bool dest_uninitialized = false) {
+    if (VM_Version::supports_avx512vlbw() && false == is_oop && MaxVectorSize  >= 32) {
```
Comment (neliasso) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2501]: false == is_oop  => !is_oop

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 9]
```diff
@@ -2095,6 +2605,10 @@ class StubGenerator: public StubCodeGenerator {
   address generate_conjoint_long_oop_copy(bool aligned, bool is_oop,
                                           address nooverlap_target, address *entry,
                                           const char *name, bool dest_uninitialized = false) {
+    if (VM_Version::supports_avx512vlbw() && false == is_oop && MaxVectorSize  >= 32) {
```
Comment (neliasso) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2608]: false == is_oop  => !is_oop

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 10]
```diff
@@ -1780,6 +2279,10 @@ class StubGenerator: public StubCodeGenerator {
   //
   address generate_disjoint_int_oop_copy(bool aligned, bool is_oop, address* entry,
                                          const char *name, bool dest_uninitialized = false) {
+    if (VM_Version::supports_avx512vlbw() && false == is_oop && MaxVectorSize  >= 32) {
```
Comment (neliasso) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:2282]: false == is_oop  => !is_oop

**Review by neliasso — CHANGES_REQUESTED** [review]
I like that you have extracted the avx512 stub code from the rest - that makes  it a lot more readable! Overall the new code feels easy to understand and read.

I found one more minor issue (appears in four places).

My only concern is that it's getting hard to follow under what circumstances avx3 instructions are used:
Could it be the case that different thresholds are needed for when you are using avx3 instructions with 32 or 64 byte vectors? Are we sure all variants are tested?

Also - have you thought about supporting oop-copies? You only have to call the BarrierSetAssembler::arraycopy_prologue/epilogue like in the old versions. It's not a requirement for me to approve this - but an encouragement for a future patch.

**Comment by jatin-bhateja** [comment]
> My only concern is that it's getting hard to follow under what circumstances avx3 instructions are used:
> Could it be the case that different thresholds are needed for when you are using avx3 instructions with 32 or 64 byte vectors? Are we sure all variants are tested?

Following 2 runtime flags influence the implementation :-
- MaxVectorSize: Determined during VM initialization based on the CPUID of the target.
- AVX3Theshold: Set to a default value of 4096 bytes based on prior performance analysis.

Following general rules were followed during implementation:
1) If target support AVX3 features (BW+VL+F) then copy will use 32 byte vectors (YMMs) for both special cases and aligned copy loop.  This is default configuration.
2) If copy length is above AVX3Threshold, then we can safely use 64 byte vectors (ZMMs) for main copy loop (and tail) since bulk of the cycles will be consumed in it.
3) Leaf level Macro Assembly routines can dynamically choose b/w YMM or ZMM register based on the AVX3Threshold value.
4) If user forces MaxVectorSize=32 then above 4096 bytes its seen that REP MOVs shows a better performance for disjoint copies. For conjoint/backward copy vector based copy performs better. 

Thus, for 32 byte vector we do not need any threshold since they execute at max frequency level.

tier1, tier2 and tier3 did not show any new issues with the changes.

> Also - have you thought about supporting oop-copies? You only have to call the
We may not see significant performance improvement considering prologue and epilogue barriers does considerable processing over object arrays.

**Comment by jatin-bhateja** [comment]
> BarrierSetAssembler::arraycopy_prologue/epilogue like in the old versions. It's not a requirement for me to approve this - but an encouragement for a future patch.

Yes that's a good pointer, I can explore extending existing GC barriers for array copy as a next step.

**Review by neliasso — APPROVED** [review]
Thanks for the clarification och update!

Reviewed.

**Comment by vnkozlov** [comment]
@jatin-bhateja Can you put summary of performance improvement into JBS?

**Comment by jatin-bhateja** [comment]
> @jatin-bhateja Can you put summary of performance improvement into JBS?



> @jatin-bhateja Can you put summary of performance improvement into JBS?

**Comment by jatin-bhateja** [comment]
> @jatin-bhateja Can you put summary of performance improvement into JBS?

Yes,  I have added the summary to JBS

**Comment by jatin-bhateja** [comment]
> @jatin-bhateja Can you put summary of performance improvement into JBS?

Hi @vnkozlov , @neliasso 
Kindly let me know your feedback, If there are no more comments is it ok to integrate this patch.

**Review by neliasso — APPROVED** [review]
Looks good to me!

**Review by vnkozlov — CHANGES_REQUESTED** [review]
The main concern which is not clear in these changes is ZMM usage which will lower frequency and case performance regression for small arrays.
That is why AVX3Threshold is set to 4096 bytes by default.
Allowing and checking for 0 AVX3Threshold  value contradicts that. Would be nice to have clear comment/explanation about that.

I also propose to use Flag constraint() functionality for checking  AVX3Threshold value instead of runtime checks everywhere. Separate RFE, please.

File: test/hotspot/jtreg/compiler/arraycopy/TestArrayCopyConjoint.java [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+package compiler.arraycopy;
+import java.util.Random;
+
+/**
+ * @test
+ * @bug 8251871
+ * @summary Optimize arrayCopy using AVX-512 masked instructions.
+ *
+ * @run main/othervm/timeout=600 -XX:-TieredCompilation  -Xbatch -XX:+IgnoreUnrecognizedVMOptions
+ *      -XX:UseAVX=3 -XX:+UnlockDiagnosticVMOptions -XX:ArrayCopyPartialInlineSize=0 -XX:MaxVectorSize=32 -XX:+UnlockDiagnosticVMOptions
```
[the remark this answers is not in the document]
Comment (jatin-bhateja) [reply test/hotspot/jtreg/compiler/arraycopy/TestArrayCopyConjoint.java:33]: Hi @vnkozlov, I have updated the pull request to cover your comments. Kindly review. 
New RFE JDK-8253721 has been created for AVX3Threshold flag related changes (PR-394).

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp [hunk 12]
```diff
@@ -1323,6 +1261,572 @@ class StubGenerator: public StubCodeGenerator {
     __ jcc(Assembler::greater, L_copy_8_bytes); // Copy trailing qwords
   }
 
+#ifndef PRODUCT
```
[the remark this answers is not in the document]
Comment (neliasso) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:1264]: But lets do that in a another change. It is good that the AVX3 case is separated out in this change - makes it easy to follow.

[the remark this answers is not in the document]
Comment (vnkozlov) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:1264]: agree

**Review by vnkozlov — APPROVED** [review]
Yes, this looks better. Reviewed. Before pushing let me test it. I will let you know results.

**Comment by vnkozlov** [comment]
hs-tier1-3 testing passed on x86 (all OSs).
